contrib/cmake/build3rdParty/build3rdParty.cpp

```diff
@@ -0,0 +1,273 @@
+void execute(string command);
+
+int main(int argc, char* argv[])
+try
+{
+    if (readCommandLine(argc, argv))
+    {
+        if (setup)
+        {
+            if (!fs::is_directory("vcpkg"))
+            {
+                execute("git clone https://github.com/microsoft/vcpkg.git");
+                execute("git clone --progress -v vcpkg vcpkg_clone");
+                fs::current_path("vcpkg");
+                #ifdef WIN32
+                    execute(".\\bootstrap-vcpkg.bat -disableMetrics");
+                #else
+                    execute("./bootstrap-vcpkg.sh -disableMetrics");
+                #endif
+            }
+            else
+            {
+                fs::current_path("vcpkg");
+            }
+
+            if (fs::exists(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_triplets" / (triplet + ".cmake")))
+            {
+                if (fs::exists(vcpkgDir / "triplets" / (triplet + ".cmake")))
+                {
+                    fs::remove(vcpkgDir / "triplets" / (triplet + ".cmake"));
+                }
+                cout << "Copying triplet from SDK: " << triplet << endl;
+                fs::copy(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_triplets" / (triplet+".cmake"), vcpkgDir / "triplets" / (triplet + ".cmake"));
+            }
+            else if (!fs::exists(vcpkgDir / "triplets" / (triplet + ".cmake")))
+            {
+                cout << "triplet not found in the SDK or in vcpkg: " << triplet << endl;
+                exit(1);
+            }
+
+            for (auto portPair : ports)
+            {
+                const string& portname = portPair.first;
+                const string& portversion = portPair.second;
+
+                if (fs::is_directory(fs::path("ports") / portname))
+                {
+                    cout << "Removing " << (vcpkgDir / "ports" / portname).u8string() << endl;
+                    fs::remove_all(vcpkgDir / "ports" / portname);
+                }
+                if (portversion.size() == 40)
+                {
+                    fs::current_path(cloneDir);
+                    execute("git checkout --quiet " + portversion);
+                    cout << "Copying port for " << portname << " from vcpkg commit " << portversion << endl;
+                    fs::copy(cloneDir / "ports" / portname, vcpkgDir / "ports" / portname, fs::copy_options::recursive);
+                    fs::current_path(vcpkgDir);
+                }
+                else
+                {
+                    cout << "Copying port for " << portname << " from SDK customized port " << portversion << endl;
+                    fs::copy(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_ports" / portname / portversion, vcpkgDir / "ports" / portname, fs::copy_options::recursive);
+                }
+            }
+
+            if (removeUnusedPorts)
+            {
+                for (auto dir = fs::directory_iterator(vcpkgDir / "ports"); dir != fs::directory_iterator(); ++dir)
```

Comment:
I believe this can be done with range-based `for`: `for (const auto& dir : fs::directory_iterator(vcpkgDir / "ports"))`

Reply:
That would be cool, I never tried it as it seemed unlikely since it doesn't have begin()/end() - if it works, does it require C++17?

Reply:
Yeah I think it works without a member begin/end! They provide [non-member overloads](https://en.cppreference.com/w/cpp/filesystem/directory_iterator/begin) for range-for compatibility

Reply:
That's great, thanks for the info.   Hopefully that is part of the pre-17 experimental implementations too - this is a good project to try that and find out.  thanks